### PR TITLE
Fix  `error_status_codes` options

### PR DESCRIPTION
**Pull request by TonyCTHsu**
**2.0 Upgrade Guide notes**

**What does this PR do?**

Fix and standardize the way to configure an array of ranges for `error_status_codes` from environment variables.

### CODE REVIEW

File: lib/datadog/tracing/contrib/status_range_matcher.rb [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+      class StatusRangeMatcher
+        def initialize(ranges)
+          @ranges = Array(ranges)
+        end
+
+        def include?(status)
+          @ranges.any? do |e|
+            case e
+            when Range
+              e.include? status
```
Comment (marcotc) [inline lib/datadog/tracing/contrib/status_range_matcher.rb:16]: How about using === for both integer and range? I think it works for both.

**Review by marcotc — APPROVED** [review]

File: lib/datadog/tracing/contrib/status_range_matcher.rb (hunk 1 above)
Comment (lloeki) [reply lib/datadog/tracing/contrib/status_range_matcher.rb:16]: Compared to current code I feel like it kind of obscures the reading though: personally I prefer reading the current version which is immediately obvious, whereas I had to check that  `2 === 2` and `(2..4) === 2` both worked.

File: lib/datadog/tracing/contrib/faraday/configuration/settings.rb [hunk 2]
```diff
@@ -42,20 +44,11 @@ class Settings < Contrib::Configuration::Settings
             option :error_status_codes do |o|
               o.env Ext::ENV_ERROR_STATUS_CODES
               o.default 400...600
```
Comment (lloeki) [inline lib/datadog/tracing/contrib/faraday/configuration/settings.rb:46]: `400...600` includes `600`, is that intended?

File: lib/datadog/tracing/contrib/excon/configuration/settings.rb [hunk 3]
```diff
@@ -39,20 +41,11 @@ class Settings < Contrib::Configuration::Settings
             option :error_status_codes do |o|
               o.env Ext::ENV_ERROR_STATUS_CODES
               o.default 400...600
```
Comment (lloeki) [inline lib/datadog/tracing/contrib/excon/configuration/settings.rb:43]: `400...600` includes `600`, is that intended?

File: lib/datadog/tracing/contrib/excon/configuration/settings.rb [hunk 4]
```diff
@@ -39,20 +41,11 @@ class Settings < Contrib::Configuration::Settings
...
-                values.map! do |v|
-                  v.gsub!(/\A[\s,]*|[\s,]*\Z/, '')
-
-                  v.empty? ? nil : v
-                end
-
-                values.compact!
-                values
+              o.setter do |v|
+                Tracing::Contrib::StatusRangeMatcher.new(v) if v
```
Comment (lloeki) [inline lib/datadog/tracing/contrib/excon/configuration/settings.rb:45]: I would add a runtime type check here, to enforce that either one of these is true:

- v is an `Integer`
- v is a `Range` with `.first` (and last, but that's internally checked by `Range`) being `Integer`

File: lib/datadog/tracing/contrib/status_range_env_parser.rb [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+                v = e.split('-')
+
+                case v.length
+                when 0 then next
+                when 1 then array << Integer(v.first)
+                when 2 then array << (Integer(v.first)..Integer(v.last))
+                else
+                  Datadog.logger.debug(
+                    "Invalid error_status_codes configuration: Unable to parse #{value}, containing #{v}."
+                  )
```
Comment (lloeki) [inline lib/datadog/tracing/contrib/status_range_env_parser.rb:23]: Somehow I feel like this logging should happen in `o.env_parser`.

Also, I think we might be inconsistent and in some places env parse fails / conf type checks raise and in others they log but IIRC we mostly raise? I might be wrong on that one but I'd rather be consistent.

If we chose to always log, I feel like we should raise `EnvValueError` / `ParseError` / `TypeError` / whatever exceptions in such pieces of code, and then the caller (the config system) of `o.env_parser` / `o.setter` / ... should handle that kind of exception uniformly and log.

**Comment by TonyCTHsu** [comment]
Going to refactor later

**Pull request merged**